messaging: support vt_message_cols to limit cols

derekperkins opened the pull request:
## Description
this enables users to set an allow-list of columns that will be loaded into memory and streamed to subscribers when they call `stream * from tbl`

## Related Issue(s)
#9666 

## Checklist
- [x] Should this PR be backported? No
- [x] Tests were added
- [ ] Documentation was added or is not required

mattlord commented:
@derekperkins I apologize for the delay, I will be reviewing this today.

derekperkins commented:
Great, thanks. I'll clean it up and get it passing tests again. I've been using this as our custom build branch trying to solve #9936. We've had this running in prod for 6 weeks or so and it has been working great.

mattlord commented on go/test/endtoend/messaging/message_test.go:
```diff
@@ -236,6 +233,75 @@ func TestThreeColMessage(t *testing.T) {
+	primary key(id),
+	index next_idx(priority, time_next desc),
+	index ack_idx(time_acked))
+comment 'vitess_message,vt_stream_cols=id;msg1,vt_ack_wait=1,vt_purge_after=3,vt_batch_size=2,vt_cache_size=10,vt_poller_interval=1'`
```
Don't feel strongly, but I think `vt_message_cols` would be better since this applies to the messages at a higher level, and at the lower level applies to both the poller and the results stream.

And beyond that ~ `createLimitedStreamingColsMessage` or ~ `TestSpecificStreamingColsMessage` would be slightly better IMO. Also minor things that I don't feel strongly about.

derekperkins replied:
done

mattlord commented on go/vt/vttablet/tabletserver/schema/load_table.go:
```diff
@@ -117,20 +102,49 @@ func loadMessageInfo(ta *Table, comment string) error {
-			continue
+	allowedStreamingCols, _ := getCols(keyvals, "vt_stream_cols")
+
+	// make sure that all the specified columns are present
```
This is where `allowed` feels misleading as you're saying that they are required. If they were allowed then I would expect that them not being present would be fine.

derekperkins replied:
renamed to `specifiedCols`

mattlord commented on go/vt/vttablet/tabletserver/schema/load_table.go:
```diff
@@ -117,20 +102,49 @@ func loadMessageInfo(ta *Table, comment string) error {
 	}
+
+	if len(allowedStreamingCols) > 0 {
+		if allowedStreamingCols[0] != "id" {
```
But I guess we don't need the other requiredCols here? 
```
		"priority",
 		"time_next",
 		"epoch",
 		"time_acked",
```

I guess those are required in the table structure but not the messages... that would make sense. This could be a reason that `vt_message_fields` is the best name...

mattlord commented on go/vt/vttablet/tabletserver/schema/load_table.go:
```diff
@@ -157,3 +171,39 @@ func getNum(in map[string]string, key string) (int, error) {
+	if sv == "" {
+		return nil, fmt.Errorf("attribute %s not specified for message table", key)
+	}
+	cols := strings.Split(sv, ";")
```
personal preference but after so many years my brain sees the semi-colon in SQL as a hard stmt/context delimiter. I can see that comma has its own problems in this context, but I think I would prefer pipe as it's not uncommon to use that as a column separator in output.

Another thing I don't feel super strongly about and leave up to you to make the final call on.

derekperkins replied:
done

mattlord commented on go/vt/vttablet/tabletserver/schema/load_table.go:
```diff
@@ -157,3 +171,39 @@ func getNum(in map[string]string, key string) (int, error) {
 	return v, nil
 }
+
+func getCols(in map[string]string, key string) ([]string, error) {
```
I feel like this should be called `parseMessageFields`?  Minor thing, I just didn't realize what it was for at first. We could also address that with a function comment.

derekperkins replied:
done

mattlord commented on go/vt/vttablet/tabletserver/schema/load_table.go:
```diff
@@ -157,3 +171,39 @@ func getNum(in map[string]string, key string) (int, error) {
+	// Load user-defined columns. Any "unrecognized" column is user-defined.
+	for _, col := range streamCols {
+		for _, field := range tableFields {
+			if field.Name == col {
```
This is case and white space sensitive, which is fine, probably just worth documenting as most identifier comparisons in SQL are not (or dependent on the collation used). We could potentially use the new evalengine here to compare the identifiers in the same way MySQL would.

mattlord replied:
To explain what I mean here, it would be something like this:
```
func getAllowedMessageFields(tableFields []*querypb.Field, streamCols []string) []*querypb.Field {
	fields := make([]*querypb.Field, 0, len(streamCols))
	// Load user-defined columns. Any "unrecognized" column is user-defined.
	for _, col := range streamCols {
		for _, field := range tableFields {
			if res, _ := evalengine.NullsafeCompare(sqltypes.NewVarChar(field.Name), sqltypes.NewVarChar(col), collations.Default()); res == 0 {
				fields = append(fields, field)
				break
			}
		}
	}
	return fields
}
```

mattlord replied:
And if we wanted to deal with the user potentially having whitespace in the comment value:
```
if res, _ := evalengine.NullsafeCompare(sqltypes.NewVarChar(strings.TrimSpace(field.Name)), sqltypes.NewVarChar(strings.TrimSpace(col)), collations.Default()); res == 0 {
				fields = append(fields, field)
				break
			}
```

derekperkins replied:
Perfect, I'll go ahead and use that. I may clean up the other functions here in a secondary commit too

mattlord replied:
Cool! To be clear, this is what I was thinking for the final function (ignoring the function name which is a separate issue):
```
func getAllowedMessageFields(tableFields []*querypb.Field, streamCols []string) []*querypb.Field {
	fields := make([]*querypb.Field, 0, len(streamCols))
	// Load user-defined columns. Any "unrecognized" column is user-defined.
	for _, col := range streamCols {
		for _, field := range tableFields {
			if res, _ := evalengine.NullsafeCompare(sqltypes.NewVarChar(field.Name),
				sqltypes.NewVarChar(strings.TrimSpace(col)),
				collations.Default()); res == 0 {
				fields = append(fields, field)
				break
			}
		}
	}
	return fields
}
```

This way the comparison is case insensitive as the default collation with MySQL 5.7 is `utf8mb4_general_ci` and with 8.0 is `utf8mb4_0900_ai_ci` and we're stripping any leading or trailing whitespace from the identifiers in the comment directive as I can see that happening (in a list you may expect whitespace not to matter).

derekperkins replied:
done

mattlord reviewed (approved):
Looks good. Thank you for working on this! This will make it feasible to have much larger message cache sizes.

I had some questions/comments/nits mostly related to naming but you can make the final call on the naming aspects. Naming things is hard, and as long as we update the docs and explain the field well that's what matters most. Speaking of that, I assume you'll have a docs PR ready to review as well? Let me know. That has gotten much easier now that we have versioned docs and I can offer help if needed.

rohit-nayak-ps reviewed (commented):
aside from minor nits, functionality lgtm

derekperkins commented on go/vt/vttablet/tabletserver/schema/load_table.go:
```diff
@@ -157,3 +171,39 @@ func getNum(in map[string]string, key string) (int, error) {
+func getCols(in map[string]string, key string) ([]string, error) {
+	sv := in[key]
+	if sv == "" {
+		return nil, fmt.Errorf("attribute %s not specified for message table", key)
```
Good call. I was just mirroring the other functions, but that was overkill. Fixed

derekperkins commented on go/vt/vttablet/tabletserver/schema/load_table_test.go:
```diff
@@ -134,6 +134,59 @@ func TestLoadTableMessage(t *testing.T) {
+	require.Equal(t, errors.New("vt_stream_cols must begin with id: test_table"), err)
+
+	// Test setting zero columns on vt_stream_cols, which is ignored and loads the default columns
+	table, err = newTestLoadTable("USER_TABLE", "vitess_message,vt_stream_cols,vt_ack_wait=30,vt_purge_after=120,vt_batch_size=1,vt_cache_size=10,vt_poller_interval=30,vt_min_backoff=10,vt_max_backoff=100", db)
```
there were already tests above without ~vt_stream_cols~`vt_message_cols` specified that I left unchanged to handle that case

derekperkins commented:
Review comments are all addressed now in the last commit in the last 2 commits. This is rebased on top of #10132. Once that is fixed and merged, I'll rebase again to get the tests passing

mattlord commented:
> Review comments are all addressed now in the last commit in the last 2 commits. This is rebased on top of #10132. Once that is fixed and merged, I'll rebase again to get the tests passing

@derekperkins Should be good now, test wise. I am still working on the locking related stuff in the other PR but the tests should be done.

derekperkins commented:
@mattlord I think this should be good to go. I ran into some rebasing issues when switching from being rebased on #10132 to main, not realizing you had squashed. In the end, it was easiest for me to just squash mine too, vs leaving the separate review commit. I addressed all the review items, via name changes and comments. May be worth a quick once over again, but no major functionality changes.

mattlord commented on go/test/endtoend/messaging/message_test.go:
```diff
@@ -274,6 +271,75 @@ func TestThreeColMessage(t *testing.T) {
 	assert.Equal(t, uint64(1), qr.RowsAffected)
 }
 
+var createSpecificStreamingColsMessage = `create table vitess_message4(
```
Any reasons not to use the new structure we have in the docs? Not a big deal, but I'd like to avoid drift between what we test and what we tell users to use if possible from here on out.

derekperkins replied:
Just didn't think to change it. I'll update the tests

derekperkins replied:
done

mattlord reviewed (approved):
Looks great! So much easier to read for me. Thank you for taking the time to make those changes, add comments, etc. It's not only easier to read, but more resilient and easier to use as well. ❤️ 


I had one minor question about the table structure used — hoping we can base it on what we now have in the docs — but that's it.

Thanks again!

mattlord commented:
Oh, one last thing... will you be opening a website PR as well so that we add details on this new configuration parameter to the v14 docs here: https://github.com/vitessio/website/blob/prod/content/en/docs/14.0/reference/features/messaging.md and here: https://github.com/vitessio/website/blob/prod/content/zh/docs/14.0/reference/messaging.md ?

I can help with that (creation or review), just let me know.

mattlord commented:
Friendly ping on this 🙂

derekperkins commented:
Working on the docs PR now
